src/c4/charconv.hpp

```diff
@@ -215,30 +215,139 @@ struct is_fixed_length
+    };
+};
+template<> struct xtoa_digits_<true, 64u>
+{
+    enum : size_t {
+        maxdigits_bin = 1 + 2 + 64, // len=67: -9223372036854775808 -0b1000000000000000000000000000000000000000000000000000000000000000
+        maxdigits_oct = 1 + 2 + 22, // len=25: -9223372036854775808 -0o1000000000000000000000
+        maxdigits_dec = 1     + 19, // len=20: -9223372036854775808 -9223372036854775808
+        maxdigits_hex = 1 + 2 + 16, // len=19: -9223372036854775808 -0x8000000000000000
+        maxdigits_bin_nopfx = 64, // len=67: -9223372036854775808 -0b1000000000000000000000000000000000000000000000000000000000000000
+        maxdigits_oct_nopfx = 22, // len=25: -9223372036854775808 -0o1000000000000000000000
+        maxdigits_dec_nopfx = 19, // len=20: -9223372036854775808 -9223372036854775808
+        maxdigits_hex_nopfx = 16, // len=19: -9223372036854775808 -0x8000000000000000
+    };
+};
+template<> struct xtoa_digits_<false, 64u>
+{
+    enum : size_t {
+        maxdigits_bin = 2 + 64, // len=66: 18446744073709551615 0b1111111111111111111111111111111111111111111111111111111111111111
+        maxdigits_oct = 2 + 22, // len=24: 18446744073709551615 0o1777777777777777777777
+        maxdigits_dec =     20, // len=20: 18446744073709551615 18446744073709551615
+        maxdigits_hex = 2 + 16, // len=18: 18446744073709551615 0xffffffffffffffff
+        maxdigits_bin_nopfx = 64, // len=66: 18446744073709551615 0b1111111111111111111111111111111111111111111111111111111111111111
+        maxdigits_oct_nopfx = 22, // len=24: 18446744073709551615 0o1777777777777777777777
+        maxdigits_dec_nopfx = 20, // len=20: 18446744073709551615 18446744073709551615
+        maxdigits_hex_nopfx = 16, // len=18: 18446744073709551615 0xffffffffffffffff
+    };
+};
+} // namespace detail
 
+// Helper macros, undefined below
 #define _c4append(c) { if(C4_LIKELY(pos < buf.len)) { buf.str[pos++] = static_cast<char>(c); } else { ++pos; } }
 #define _c4appendhex(i) { if(C4_LIKELY(pos < buf.len)) { buf.str[pos++] = hexchars[i]; } else { ++pos; } }
 
-}
-#include <iostream>
-namespace c4 {
 C4_INLINE_CONSTEXPR const char hexchars[] = "0123456789abcdef";
 
+
 /** write an integer to a string in decimal format. This is the
  * lowest level (and the fastest) function to do this task.
  * @note does not accept negative numbers
  * @return the number of characters required for the string,
  * even if the string is not long enough for the result.
  * No writes are done past the end of the string. */
 template<class T>
-size_t write_dec(substr buf, T v)
+size_t write_dec(substr buf, T v) noexcept
 {
     C4_STATIC_ASSERT(std::is_integral<T>::value);
     C4_ASSERT(v >= 0);
+    if(C4_UNLIKELY(buf.len < detail::xtoa_digits<T>::maxdigits_dec))
+        return detail::xtoa_digits<T>::maxdigits_dec;
     size_t pos = 0;
     do {
-        _c4append('0' + (v % T(10)));
+        buf.str[pos++] = (char)('0' + (v % T(10)));
```

Comment:
> How did the check on buffer length made a 10/20% enhancement ? is the benchmark being fed with incompatible numbers (too large for buffer) ? meaning when you benchmark you do only with nominal cases, not erroneous ones right ?

@fargies the `_c4append()` which was used in this place before is an if, now removed. So instead of `pos` checks there is only one. So less code instructions and less work for branch prediction. And indeed the benchmark results are for the happy case. The buffer-too-small case will see larger improvements, but I'm not really worried about that case.

Reply:
Makes sense, thanks